source/simulator/entropy_viscosity.cc

```diff
@@ -0,0 +1,439 @@
+
+        if (advection_field.is_temperature())
+          {
+            scratch.finite_element_values[solution_field].get_function_laplacians (old_solution,
+                                                                                   scratch.old_field_laplacians);
+            scratch.finite_element_values[solution_field].get_function_laplacians (old_old_solution,
+                                                                                   scratch.old_old_field_laplacians);
+          }
+
+        if (parameters.include_melt_transport && melt_handler->is_porosity(advection_field))
+          {
+            scratch.finite_element_values[introspection.extractors.velocities].get_function_divergences (current_linearization_point,
+                scratch.current_velocity_divergences);
+          }
+
+        /**
+         * Explicit material model inputs and outputs.
+         */
+        for (unsigned int q=0; q<n_q_points; ++q)
+          {
+            scratch.material_model_inputs.temperature[q] = (scratch.old_temperature_values[q] + scratch.old_old_temperature_values[q]) / 2;
+            scratch.material_model_inputs.position[q] = scratch.finite_element_values.quadrature_point(q);
+            scratch.material_model_inputs.pressure[q] = (scratch.old_pressure[q] + scratch.old_old_pressure[q]) / 2;
+            scratch.material_model_inputs.velocity[q] = (scratch.old_velocity_values[q] + scratch.old_old_velocity_values[q]) / 2;
+            scratch.material_model_inputs.pressure_gradient[q] = (scratch.old_pressure_gradients[q] + scratch.old_old_pressure_gradients[q]) / 2;
+
+            for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+              scratch.material_model_inputs.composition[q][c] = (scratch.old_composition_values[c][q] + scratch.old_old_composition_values[c][q]) / 2;
+            scratch.material_model_inputs.strain_rate[q] = (scratch.old_strain_rates[q] + scratch.old_old_strain_rates[q]) / 2;
+          }
+        scratch.material_model_inputs.cell = &cell;
+        create_additional_material_model_outputs(scratch.material_model_outputs);
+
+        material_model->evaluate(scratch.material_model_inputs,scratch.material_model_outputs);
+        MaterialModel::MaterialAveraging::average (parameters.material_averaging,
+                                                   cell,
+                                                   scratch.finite_element_values.get_quadrature(),
+                                                   scratch.finite_element_values.get_mapping(),
+                                                   scratch.material_model_outputs);
+
+        if (parameters.formulation_temperature_equation ==
+            Parameters<dim>::FormulationTemperatureEquation::reference_density_profile)
+          for (unsigned int q=0; q<n_q_points; ++q)
+            scratch.material_model_outputs.densities[q] = adiabatic_conditions->density(scratch.material_model_inputs.position[q]);
```

Comment:
I think you should probably do this loop before the call to `MaterialAveraging::average` above. The same may be true in other assemblers but I didn't pay attention there -- please double check that the averaging always happens *after* the correct densities have been computed.